analyzers/src/SonarAnalyzer.CSharp/Rules/FrameworkViewCompiler.cs

```diff
@@ -0,0 +1,339 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.IO;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Xml.Linq;
+using System.Xml.XPath;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class FrameworkViewCompiler : SonarDiagnosticAnalyzer
+{
+    // TODO: Read this from context.ProjectConfiguration().ProjectPath.
+    // Not sure how to set it up in the UTs.
+    private const string ROOT_DIR = """C:\dev\personal\aspnet\Lol""";
+
+    private ImmutableArray<DiagnosticAnalyzer> Rules;
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+    protected override bool EnableConcurrentExecution => false;
+
+    public FrameworkViewCompiler()
+    {
+        Rules = RuleFinder2.CreateAnalyzers(AnalyzerLanguage.CSharp, false).ToImmutableArray();
+        SupportedDiagnostics = Rules.SelectMany(x => x.SupportedDiagnostics).ToImmutableArray();
+    }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+
+        context.RegisterCompilationAction(
+            c =>
+            {
+                var mightBeUseful = c.ProjectConfiguration();
+                var supportedDiagnostics = SupportedDiagnostics.ToImmutableDictionary(x => x.Id, x => ReportDiagnostic.Warn);
+
+                var dummy = CompileViews(c.Compilation)
+                    .WithOptions(c.Compilation.Options.WithSpecificDiagnosticOptions(supportedDiagnostics))
+                    .WithAnalyzers(Rules);
+
+                var diagnostics = dummy.GetAllDiagnosticsAsync().Result;
+                foreach (var diagnostic in diagnostics.Where(x => supportedDiagnostics.Keys.Contains(x.Id)))
+                {
+                    // TODO: This breaks when raising on out-of-compilation files. Find a replacement, maybe writing directly to the SARIF.
+                    c.ReportIssue(diagnostic);
```

Comment:
See [this thread](https://sonarsource.slack.com/archives/C05E77TRN1H/p1714643664768389).

The main problem is that:

- On .NET Core, the Views _are_ part of the compilation...but we do not care, as they are compiled anyways and we get analysis (almost) for free.
- On .NET Framework the Views are NOT part of the compilation, so if we propagate the diagnostic with `ReportIssue`, roslyn throws an `AD0001`.